testdriver/testdriver.go

```diff
@@ -538,46 +515,145 @@ func (d *TestDriver) listenDriver() error {
+		fmt.Fprintln(os.Stderr, resp.vimError)
 	}
+	if resp.value != nil {
+		v := *resp.value
+		if *bang {
+			ef("unexpected command success")
+		}
+		if *stringout {
+			switch v := v.(type) {
+			case string:
+				fmt.Print(v)
+			default:
+				ef("response type is %T, not string", v)
+			}
+		} else {
+			enc := json.NewEncoder(os.Stdout)
+			if *indent {
+				enc.SetIndent("", "  ")
+			}
+			if err := enc.Encode(v); err != nil {
+				ef("failed to format output of JSON: %v", err)
+			}
+		}
+	}
+	return 0
+}
+
+func VimExprWait() (exitCode int) {
+	defer cleanUp(&exitCode)
 
 	fs := flag.NewFlagSet("vim", flag.PanicOnError)
-	bang := fs.Bool("bang", false, "expect command to fail")
-	indent := fs.Bool("indent", false, "pretty indent resulting JSON")
+	noindent := fs.Bool("noindent", false, "do not pretty indent resulting JSON")
 	stringout := fs.Bool("stringout", false, "print resulting string rather than JSON encoded version of string")
-
-	log("starting vim driver client and parsing flags...")
+	fWait := fs.String("wait", "", "retry (with exp backoff) until this time period has elapsed")
 
 	fs.Parse(os.Args[1:])
 	args := fs.Args()
+	if len(args) < 2 {
+		ef("invalid arguments")
+	}
+	fileName := args[0]
+	want, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		ef("failed to read %v: %v", fileName, err)
+	}
+	// Now repeatedly evaluate the provided expression
+	// with exp backoff until we match want. Abort on any errors
+	// because this might indicative of a greater problem.
+	vimCmd := append([]string{"expr"}, args[1:]...)
+	if *fWait == "" {
+		fWait = &DefaultErrLogMatchWait
+	}
+	wait, err := time.ParseDuration(*fWait)
+	if err != nil {
+		ef("failed to parse -wait duration %q: %v", *fWait, err)
+	}
+	strategy := retry.LimitTime(wait,
+		retry.Exponential{
+			Initial: 100 * time.Millisecond, // be slightly less aggressive with Vim
+			Factor:  1.5,
+		},
+	)
+	var got *bytes.Buffer
+	for a := retry.Start(strategy, nil); a.Next(); {
+		resp, err := vim(vimCmd)
+		if err != nil {
+			ef("%v", err)
+		}
+		if resp.protocolError != "" {
+			ef("protocol error: %v", resp.protocolError)
+		}
+		if resp.vimError != "" {
+			ef("vim error: %v", resp.vimError)
+		}
+		if resp.value == nil {
+			ef("expected return value from Vim; didn't get one")
+		}
+		v := *resp.value
+		got = new(bytes.Buffer)
+		if *stringout {
+			switch v := v.(type) {
+			case string:
+				got.WriteString(v)
+			default:
+				ef("response type is %T, not string", v)
+			}
+		} else {
+			enc := json.NewEncoder(got)
+			if !*noindent {
+				enc.SetIndent("", "  ")
+			}
+			if err := enc.Encode(v); err != nil {
+				ef("failed to format output of JSON: %v", err)
+			}
+		}
+		if bytes.Equal(want, got.Bytes()) {
+			return 0
+		}
+	}
+	// We failed to find a match. Show diff and return an error
+	fmt.Fprint(os.Stderr, textutil.Diff(got.String(), string(want)))
+
+	return 1
+}
+
+type vimResponse struct {
+	protocolError string
+	vimError      string
```

Comment:
Optional, but I'd make this a proper error, for readability and so it can be an arbitrary value.

Reply:
Thanks - yep with you on this too.